### PR TITLE
Logical short circuit

### PR DESCRIPTION
**Pull request by sundy-li**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- Improvement

Short description (up to few sentences):

Add simple logical short circuit optimization.

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
`allow_shORt_circuit_logic_expressions`
It's too hacky, maybe name it "allow_short_circuit_or"?

**Comment by sundy-li** [comment]
Yes, it's just accidental search-and-replace. I have considered other names like `allow_constant_folding_logic_expressions`,  `allow_logic_expressions_constant_folding`

**Comment by alexey-milovidov** [comment]
> I have considered other names like allow_constant_folding_logic_expressions, allow_logic_expressions_constant_folding

Ok.

File: dbms/src/Interpreters/Settings.h [hunk 1]
```diff
@@ -306,6 +306,7 @@ struct Settings
     M(SettingBool, cancel_http_readonly_queries_on_client_close, false, "Cancel HTTP readonly queries when a client closes the connection without waiting for response.") \
     M(SettingBool, external_table_functions_use_nulls, true, "If it is set to true, external table functions will implicitly use Nullable type if needed. Otherwise NULLs will be substituted with default values. Currently supported only for 'mysql' table function.") \
     M(SettingBool, allow_experimental_data_skipping_indices, false, "If it is set to true, data skipping indices can be used in CREATE TABLE/ALTER TABLE queries.")\
+    M(SettingBool, allow_short_circuit_logic_expressions, false, "If it is set to true, logic expressions will be optimized by short circuit, but this may make the alias lost by optimization")\
```
Comment (alexey-milovidov) [inline dbms/src/Interpreters/Settings.h]: Lost alias shouldn't be an option. We can push aliases to `WITH` section of the query.

Comment (sundy-li) [reply dbms/src/Interpreters/Settings.h]: `select a  from table where 1 = 1 or  (( device like '%android%' ) as t)`
will be optimized to 
`select a  from table where 1`,  the device column will never be read,  should we keep the alias `t` ?

Comment (alexey-milovidov) [reply dbms/src/Interpreters/Settings.h]: If you write
```
with ( device like '%android%' ) as t select a from table where 1
```
column `device` will not read.

File: dbms/src/Interpreters/LogicalExpressionsOptimizer.cpp [hunk 2]
```diff
@@ -77,6 +91,125 @@ void LogicalExpressionsOptimizer::perform()
...
+            bool is_or = function->name == "or";
+            auto * args = typeid_cast<ASTExpressionList *>(function->arguments.get());
+            for (auto it = args->children.begin(); it != args->children.end();)
+            {
+                auto value = tryExtractAndReplaceConstColumn(*it);
+                if (value.second)
+                {
+                    bool flag = value.first.column->getBool(0);
+                    /// TODO: deal with the alias, but here we just ignore the alais by Short circuit evaluation
+                    if ((flag && is_or) || (!flag && !is_or))
```
Comment (vitlibar) [inline dbms/src/Interpreters/LogicalExpressionsOptimizer.cpp]: // Comment with example?
`if (flag == is_or)`

File: dbms/src/Interpreters/LogicalExpressionsOptimizer.cpp [hunk 3]
```diff
@@ -77,6 +91,125 @@ void LogicalExpressionsOptimizer::perform()
...
+            ColumnNumbers args_position;
+            size_t position = 0;
+            auto * func_args = typeid_cast<ASTExpressionList *>(function->arguments.get());
+
+            bool flag = true;
+            for (auto & child : func_args->children)
+            {
+                auto value = tryExtractAndReplaceConstColumn(child);
+                if (!value.second)
+                    flag = false;
```
Comment (vitlibar) [inline dbms/src/Interpreters/LogicalExpressionsOptimizer.cpp]: What does this `flag` mean? Why it's `false` here?

File: dbms/tests/queries/0_stateless/00908_analyze_query.reference [hunk 4]
```diff
@@ -1 +1,23 @@
...
+SELECT number\nFROM numbers(10) \nWHERE rand() > 4
+SELECT number\nFROM numbers(10) \nWHERE 1 AS a
+SELECT \n    rand(), \n    rand(), \n    rand() + rand()
+SELECT 33
+SELECT if(number = 44, 33, 44)\nFROM numbers(10) 
+SELECT number\nFROM numbers(10) \nWHERE 1
+SELECT number\nFROM numbers(10) \nWHERE 0
+SELECT number\nFROM numbers(10) \nPREWHERE 0
+SELECT number\nFROM numbers(10) \nWHERE 0
+<<<<<<< HEAD
```
Comment (abyss7) [inline dbms/tests/queries/0_stateless/00908_analyze_query.reference]: You have a conflict marker committed

**Comment by sundy-li** [comment]
Sorry, I will make another PR later.

**Pull request closed without merging**